packages/appium/lib/appium.js

```diff
@@ -241,6 +251,261 @@ class AppiumDriver extends DriverCore {
+        // Not sure if we need to do anything here if the client closes the websocket connection.
+        // Probably if a session was started via the socket, and the socket closes, we should end the
+        // associated session to free up resources. But otherwise, for sockets attached to existing
+        // sessions, doing nothing is probably right.
+        this.log.debug(`Bidi socket connection closed (code ${code}, reason: '${reason}')`);
+
+        // If we're proxying, might as well close the upstream connection and clean it up
+        if (proxyClient) {
+          this.log.debug('Also closing bidi proxy socket connection');
+          proxyClient.close(code, reason);
+        }
+      });
+
+      // If the driver emits a bidi event that should maybe get sent to the client, check to make
+      // sure the client is subscribed and then pass it on
+      let eventListener = async ({context, method, params}) => {
+        // if the driver didn't specify a context, use the empty context
+        if (!context) {
+          context = '';
+        }
+        if (!method || !params) {
+          throw new Error(
+            `Driver emitted a bidi event that was malformed. Require method and params keys ` +
+              `(with optional context). But instead received: ${JSON.stringify({
+                context,
+                method,
+                params,
+              })}`,
+          );
+        }
+        if (ws.readyState !== 1) {
+          // if the websocket is not still 'open', then we can ignore sending these events
+          if (ws.readyState > 1) {
+            // if the websocket is closed or closing, we can remove this listener as well to avoid
+            // leaks
+            bidiHandlerDriver.eventEmitter.removeListener(BIDI_EVENT_NAME, eventListener);
+          }
+          return;
+        }
+
+        if (
+          bidiHandlerDriver.bidiEventSubs[method] &&
+          bidiHandlerDriver.bidiEventSubs[method].includes(context)
+        ) {
+          this.log.info(
+            `<-- BIDI EVENT ${method} (context: '${context}', params: ${JSON.stringify(params)})`,
+          );
+          // now we can send the event onto the socket
+          const ev = {type: 'event', context, method, params};
+          // TODO add conditional logic to send only if user is subscribed
+          await send(JSON.stringify(ev));
+        }
+      };
+      bidiHandlerDriver.eventEmitter.on(BIDI_EVENT_NAME, eventListener);
+    } catch (err) {
+      this.log.error(err);
+      try {
+        ws.close();
+      } catch (ign) {}
+    }
+  }
+
+  /**
+   * @param {Buffer} data
+   * @param {ExternalDriver | AppiumDriver} driver
+   */
+  async onBidiMessage(data, driver) {
+    let resMessage, id, method, params;
+    try {
+      try {
+        ({id, method, params} = JSON.parse(data.toString('utf8')));
+      } catch (err) {
+        throw new errors.InvalidArgumentError('Could not parse Bidi command');
```

Comment:
it makes sense to also print the command and the error itself

Reply:
I would still prefer something like `Could not parse Bidi command '${_.truncate(data ...)}': ${err.message}`